### ISSUE TITLE
Protect application from DDOS or loss of service

### ISSUE THREAD

**Issue by Jouster500** [megascatterbomb/MegaAntiCheat#30]
The application as proposed is almost certainly prone to bad actors. Given that the plan is to provide players with a client that will in principle make calls back to the database is open for said bad actors to reverse engineer the api calls and have a means to make calls to database without the client itself. This in turn allows bad actors to automate thousands of calls to the database with noise, preventing legitimate requests from being processed, effectively shutting down the service; No database means no lookups for cheaters which cheaters ingame can then act with impunity. 

Investigate methods to mitigate these DDOS attempts, and provide a way to get the database back online in a sufficient time frame in the event of outage.

**Comment by AnixAnalysis** [comment]
> nothing firewalling and rate limiting shouldn't be able to handle 🙂

I agree on some levels but if/when we're getting slammed with a DDOS the size 1000+ bot machines we could just level the whole server on rate limiting alone.

Depending on the budget you could employ the following:

- Putting the API behind cloud flare - not great for APIs but better then nothing and will provide some bare level DDOS protection.
- Speaking with the provider (i believe you're using OVH, if not please correct me!) to see if you can get an enchanced DDOS protection (would probabbly cost money). I'm sure OVH has some level of DDOS protection built in but i hear its not perfect, never had to use it myself.
- Employing multiple smaller servers served behind a loadbalancer all connected to the database backend. More cost but depending on how many servers we could scale out or scale up when needed but will be slow.
- Going cloud, aka the above but more automated... Gonna be hella costly if it gets hit but will handle most things within minutes depending on detection and scaling settings.

Please correct me or add onto this if you feel there is another option i am missing.

(I am a devops/operations side person for a sizable company that uses azure, but i muck around with OVH and bare metal to run my own stuff)

**Comment by Defiect** [comment]
Link to the relevant discussion thread for this topic #31

**Comment by TheArcaneBrony** [comment]
This is why I'd proposed making the service federated on Discord.

**Comment by davidhatten** [comment]
A rate limiting library will be a must, there's no reason for someone to make 10 requests in a minute. Or even 1 request every minute for 30 minutes. There are behaviors that can be set to lock out based on IP and other factors. You can throttle authenticated traffic that looks suspicious too. Unauthenticated traffic can be rate limited but that shouldn't be able to get far enough to do damage. Which brings me to my second point....

It might be worth making this entire application asynchronous. Client sends request to AntiCheat. AntiCheat controller authenticates and shoves the data into a background worker (assuming auth passes), doing no real work itself. Meanwhile, the controller responds `202 - accepted` and the client hangs around waiting to receive a callback if it needs a response from the server based on the work that's being done.

it's a bit more complicated, but it gives you more layers of abstraction to implement security and separates your server thread from any work that could block it.

**Comment by TheArcaneBrony** [comment]
@davidhatten I think your rate limits are far too low, you would be blocking clients attempting to make legitimate queries here

**Comment by davidhatten** [comment]
The specific numbers will come out with true use case discussion and observation. If a client sends the exact same request 10x in a minute with the same data, then that's sus. But 10 requests in a minute each for a different steamID is (probably) totally legit traffic and should be let through. But again, the point is less the numbers and more "Have Rate Limiting With Solid Rules"

**Comment by Chopper1337** [comment]
Try heavy rate limiting and a local cached version of the database for more frequent access. The local version could be synced at some rate.

**Comment by davidhatten** [comment]
Downside to async background worker model: Doing it in a way that results in a lossless application will result in a 2nd set of hosts to hold to job info. My experience is with an application backed with redis, but the point is that the library uses the keystore to store the job queue so even if the application goes down, the work that was enqueued can be worked on when the application recovers.
